lightning/src/util/persist.rs

```diff
@@ -250,3 +261,942 @@ where
+/// hundred bytes.
+///
+/// Note that monitors written with this struct are _not_ backward-compatible
+/// with the default [`KVStore`]'s [`Persist`] implementation. They have a
+/// prepended byte sequence, [`MONITOR_UPDATING_PERSISTER_PREPEND_SENTINEL`],
+/// applied to prevent deserialization with other persisters. This is because
+/// monitors written by this struct _may_ have unapplied updates, particularly
+/// if the node panics and aborts. In order to downgrade, you must ensure that
+/// all updates are applied to the monitor, and remove the sentinel bytes. For
+/// more details, please see the release notes for 0.0.117.
+///
+/// # Storing monitors
+///
+/// Monitors are stored by implementing the [`Persist`] trait, which has two
+/// functions:
+///
+///   - [`persist_new_channel`], which persists whole [`ChannelMonitor`]s.
+///   - [`update_persisted_channel`], which persists only a
+///  [`ChannelMonitorUpdate`]
+///
+/// Whole [`ChannelMonitor`]s are stored in the
+/// [`CHANNEL_MONITOR_PERSISTENCE_NAMESPACE`], using the familiar encoding of an
+/// [`OutPoint`]'s transaction ID and index (for example,
+/// `[SOME-64-CHAR-HEX-STRING]_1`).
+///
+/// Each [`ChannelMonitorUpdate`] is stored in a dynamic sub-namespace, as follows:
+///
+///   - namespace: [`CHANNEL_MONITOR_UPDATE_PERSISTENCE_NAMESPACE`]
+///   - sub-namespace: [the monitor's encoded OutPoint name]
+///
+/// Under that namespace, each update is stored with a left-padded `u64`, like
+/// `00000000000000000021`.
+///
+/// For example, consider this channel:
+///
+///   - Outpoint
+///   - Transaction ID:
+///     `deadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33f`
+///   - Index: `1`
+///
+/// Full channel monitors would be stored at a single key:
+///
+/// `[CHANNEL_MONITOR_PERSISTENCE_NAMESPACE]/deadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33f_1`
+///
+/// Updates would be stored as follows (with `/` delimiting
+/// namespace/subnamespace/key):
+///
+/// `[CHANNEL_MONITOR_UPDATE_PERSISTENCE_NAMESPACE]/deadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33f_1/00000000000000000002`
+/// `[CHANNEL_MONITOR_UPDATE_PERSISTENCE_NAMESPACE]/deadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33f_1/00000000000000000003`
+/// `[CHANNEL_MONITOR_UPDATE_PERSISTENCE_NAMESPACE]/deadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33f_1/00000000000000000004`
+/// ... and so on.
+///
+/// # Reading channel state from storage
+///
+/// Channel state is reconstructed by calling
+/// [`read_channel_monitors_with_updates`]. Please see that function's
+/// documentation for more information.
+///
+/// # Pruning stale channel updates
+///
+/// TODO(domz)
+pub struct MonitorUpdatingPersister<K: Deref, L: Deref>
+where
+	K::Target: KVStore,
+	L::Target: Logger,
+{
+	kv_store: K,
+	logger: L,
+	maximum_pending_updates: u64,
+	last_monitor_update_id: RefCell<HashMap<MonitorName, u64>>,
```

Comment:
Yeah, I had this concern too, and considered a `Mutex`, and almost just did it because the inefficiency would be miniscule relative to I/O. Wanted to ask LDK for some guidance here though. If there's doubt, `Mutex` is probably better.

Reply:
I pushed an update that uses the on-disk CMs as the "source of truth" for the last full CM update_id, with maximum updates implemented via a modulo calculation, per suggestion from @G8XSU. That obviates this hashmap. 

Noted that if we introduce anything else like this (or revert that change), it should be mutexed.